cmd/root.go

```diff
@@ -50,96 +53,209 @@ const (
+// and error, etc. In practice, most of it is normally accessed through the `os`
+// package from the Go stdlib.
+//
+// We group them here so we can prevent direct access to them from the rest of
+// the k6 codebase. This gives us the ability to mock them and have robust and
+// easy-to-write integration-like tests to check the k6 end-to-end behavior in
+// any simulated conditions.
+//
+// `newGlobalState()` returns a globalState object with the real `os`
+// parameters, while `newGlobalTestState()` can be used in tests to create
+// simulated environments.
+type globalState struct {
+	ctx context.Context
+
+	fs      afero.Fs
+	getwd   func() (string, error)
+	args    []string
+	envVars map[string]string
+
+	defaultFlags, flags globalFlags
+
+	outMutex       *sync.Mutex
+	stdOut, stdErr *consoleWriter
+	stdIn          io.Reader
+
+	signalNotify func(chan<- os.Signal, ...os.Signal)
+	signalStop   func(chan<- os.Signal)
+
+	// TODO: add os.Exit()?
 
+	logger         *logrus.Logger
+	fallbackLogger logrus.FieldLogger
+}
+
+// Ideally, this should be the only function in the whole codebase where we use
+// global variables and functions from the os package. Anywhere else, things
+// like os.Stdout, os.Stderr, os.Stdin, os.Getenv(), etc. should be removed and
+// the respective properties of globalState used instead.
+func newGlobalState(ctx context.Context) *globalState {
 	isDumbTerm := os.Getenv("TERM") == "dumb"
 	stdoutTTY := !isDumbTerm && (isatty.IsTerminal(os.Stdout.Fd()) || isatty.IsCygwinTerminal(os.Stdout.Fd()))
 	stderrTTY := !isDumbTerm && (isatty.IsTerminal(os.Stderr.Fd()) || isatty.IsCygwinTerminal(os.Stderr.Fd()))
 	outMutex := &sync.Mutex{}
-	return &commandFlags{
-		defaultConfigFilePath: defaultConfigFilePath,  // Updated with the user's config folder in the init() function below
-		configFilePath:        os.Getenv("K6_CONFIG"), // Overridden by `-c`/`--config` flag!
-		exitOnRunning:         os.Getenv("K6_EXIT_ON_RUNNING") != "",
-		showCloudLogs:         true,
-		runType:               os.Getenv("K6_TYPE"),
-		archiveOut:            "archive.tar",
-		outMutex:              outMutex,
-		stdoutTTY:             stdoutTTY,
-		stderrTTY:             stderrTTY,
-		stdout:                &consoleWriter{colorable.NewColorableStdout(), stdoutTTY, outMutex, nil},
-		stderr:                &consoleWriter{colorable.NewColorableStderr(), stderrTTY, outMutex, nil},
+	stdOut := &consoleWriter{os.Stdout, colorable.NewColorable(os.Stdout), stdoutTTY, outMutex, nil}
+	stdErr := &consoleWriter{os.Stderr, colorable.NewColorable(os.Stderr), stderrTTY, outMutex, nil}
+
+	logger := &logrus.Logger{
+		Out: stdErr,
+		Formatter: &logrus.TextFormatter{
+			ForceColors: stderrTTY,
+			// This is a hack just for this env var, since we need the logger
+			// before we can parse all of the env vars...
+			DisableColors: !stderrTTY || os.Getenv("NO_COLOR") != "" || os.Getenv("K6_NO_COLOR") != "",
```

Comment:
Doesn't this need ot check that `NO_COLOR` is set at all ?

Reply:
:thinking: Maybe, but I am somewhat reluctant to do it that way, doesn't seem super safe or practical... It's as easy to set `NO_COLOR=true` as it is to set `NO_COLOR=`, right? But in the reverse case, if you had a global `NO_COLOR=true` variable, it's not as easy to unset it. If we accepted empty strings values, you'd need to do this:
```sh
unset NO_COLOR
k6 run script.js
```

whereas with the current code, you can do this:
```
NO_COLOR="" k6 run script.js
```

Reply:
Maybe,  but we both do it like that in the other place you check this *and* (probably more importantly) this is how `NO_COLOR` is defined to work "if it exists" not depending on it's value. 

Given that whoever will be using k6 is likely to set  `NO_COLOR` globally in order to disable color in everything, it will be kind of bad of us to not actually do the same thing as the other programs supporting https://no-color.org/ :shrug: . 


I really doubt there is a practical case where somebody having disabled colors *globally* will want to specifically enable them for us. 

Also, this is how it actually works because you handle `NO_COLOR` in another place as well. This only matters for early errors like 
![image](https://user-images.githubusercontent.com/312246/158587797-4955610f-22d4-4b38-94c9-69e76c05bbfe.png)
Also note the format change :thinking:

Reply:
I looked at some of the projects listed on https://no-color.org/ and they do exactly what I do here, check if NO_COLOR actually has a value. See [this](https://github.com/muesli/termenv/blob/0899c83a5e78cfd47f3077b860dc4a94f54ce55b/termenv.go#L77-L84) and [this](https://github.com/xyproto/textoutput/blob/dec1ab3255b039faa9afd059be44afa9026936e4/textoutput.go#L37) for example. So I don't think I should change it, I think the "spec" is just poorly written :man_shrugging: 

Regarding the format change, I think it's is just something `logurs.TextFormatter` does :thinking: In any case, if we decide to fix it, I think that should be a separate PR, this behavior is the same for the one we have for non-TTY terminals right now and maybe both need to be fixed :man_shrugging:

Reply:
>  Command-line software which adds ANSI color to its output by default should check for the presence of a NO_COLOR environment variable that, when present (regardless of its value), prevents the addition of ANSI color.

Reply:
Also, *again* this PR *already* disables colors if NO_COLOR is set to an empty value 


![image](https://user-images.githubusercontent.com/312246/158591420-17116494-25b7-40da-a4ce-9b49d593226f.png)
Just that happens a bit later, so it *doesn't* get disabled if k6 errors out early as I show in the previous screenshot. See also @oleiade [comment](https://github.com/grafana/k6/pull/2410#issuecomment-1060770737). 

This happens in [getFlags](https://github.com/grafana/k6/pull/2410/files#diff-ab967ab1a2f3a1b769106eeb7bfe892ef0e81d1d27811fa15be08e6749feee1fR187-R191)

Reply:
As I mention I kind of prefer the `NO_COLOR= command` disabling the colors as I said. WDYT @oleiade @codebien @olegbespalov 

I am fine with this being removed and no actually add `NO_COLOR` if we do not agree on how it should behave

Reply:
hmm 3rd options: `NO_COLOR` will work without a value, but `K6_NO_COLOR` requires one? :sweat_smile:

Reply:
Resolved by the latest [force-push](https://github.com/grafana/k6/compare/ca4cc570c4a9de37022244dbae704e4fb3def926..d49634c76b9cabc355b79c167581534c85e753b7)?

Reply:
I'm not entirely sure if I understood what the different visions in this discussion are.

Thus, reading the https://no-color.org specs, and judging by how my favorite shell (fish), albeit a bit unconventional, handles this case, my preferred option is to have this behavior (which, I believe, is what this PR does as of now?):

```
# Colorful output
k6 run script.js

# No color output (set for the execution of this command only, the next invocation without the `NO_COLOR=` statement would be colorful)
NO_COLOR= k6 run script.js

# No color output (set for the execution of this command only, the next invocation without the `NO_COLOR=` statement would be colorful)
K6_NO_COLOR= k6 run script.js

# persistent no color (fish syntax)
set -x NO_COLOR
k6 run script.js  # no colors
k6 run script.js  # still no colors
set -e NO_COLOR
k6 run script.js # colorful
```

**N.B** interestingly, I noticed that fish shell has a different behavior (in appearance), compared to bash, when it comes to empty env variable:
```fish
> set -x NO_COLOR
> env | grep NO_COLOR
NO_COLOR=
```

Reply:
> # No color output (set for the execution of this command only, the next invocation without the `NO_COLOR=` statement would be colorful)
> K6_NO_COLOR= k6 run script.js

This will now be colorful with the current changes. But `K6_NO_COLOR=literallyanything` will including `K6_NO_COLOR=false`